HW0/P6/P6.txt

```diff
@@ -1 +1,41 @@
-# Your answers here
+###########################################################################################
+## Problem 6A
+###########################################################################################
+
+When running P6A.py, the 10 jobs did not always finish running in the same order that they started. For example, the code output below shows that the first batch of four jobs started in order 0, 1, 2, 3 but job 1 finished before job 0. Similarly, the second batch of four jobs started in order 4, 5, 6, 7 but job 7 finished before job 6.
+
+This appears to be 'normal' behavior rather than an outlier; in fact, the code had to be run 15 times before getting an output where all 10 jobs ended in the same order that they had started. This implies that it is difficult to guarantee consistent runtimes across individual jobs, even for very simple jobs like this one, and that it is likely to be even harder with larger jobs.
+
+To test this, I increased the sleep time from 0.25 to 1.0.  As expected, this resulted in the jobs being even more out of order with each other: in a few cases, the second or third batches of jobs started running before all the jobs in the previous batch had finished running, and in once instance even the first batch of jobs started out of order. In fact, out of 30 total runs, the jobs did not once end in the same order that they had started. These results indicate that smaller job sizes are necessary when attempting to maintain consistent run times across jobs and/or ordered jobs.
+
+One scenario where this type of behavior would be important is the case where subsequent jobs require inputs from earlier jobs. Such dependencies could result in bottlenecks, where processors are idle while waiting from other jobs to finish running. If not handled correctly, this situation could also result in later jobs using the wrong inputs (for example, if they are not able to detect that the earlier job has not finished running and use garbage values as input). This problem could be amplified in cases where larger jobs are unavoidable - for example, to minimize or offset the overall time associated with transferring a large dataset that is necessary for each job - as they could end up being run even more out of order.
```

Comment:
I think that having the bottlenecks is inevitable if you want to ensure that later jobs is using the right values. This is the concept of Locks on shared data, so that only 1 thread can access a shared piece of data at any time. Bottlenecks are much better than incorrect input values!